src/actions/attack.cpp

```diff
@@ -1367,6 +1367,25 @@ void attack::perform()
 		return;
 	}
 
+	bc_.reset(new battle_context(units_, a_.loc_, d_.loc_, a_.weapon_, d_.weapon_));
+
+	a_stats_ = &bc_->get_attacker_stats();
+	d_stats_ = &bc_->get_defender_stats();
+
+	if(a_stats_->weapon) {
+		a_.weap_id_ = a_stats_->weapon->id();
+	}
+
+	if(d_stats_->weapon) {
+		d_.weap_id_ = d_stats_->weapon->id();
+	}
+
+	try {
+		fire_event("pre_attack");
+	} catch(const attack_end_exception&) {
+		return;
+	}
+
 	a_.get_unit().set_facing(a_.loc_.get_relative_dir(d_.loc_));
 	d_.get_unit().set_facing(d_.loc_.get_relative_dir(a_.loc_));
```

Comment:
I'm not sure of the purpose of the `pre_attack` event, does it make sense for the units to face each other first, and to face each other even if an `attack_end_exception` happens?

Setting it first would allow the event to be sure that the units' facing already matched the attack direction. I guess setting it afterwards might allow a "charge bonus" if the unit's movement already made it face the correct direction.

Reply:
The purpose of it is described in the FR issue.

I don't understand the significance of the direction the units are facing? My immediate thought would be it would make sense for them to be facing each other before the event is fired, since the decision to attack has already been made, even if it ends up not actually happening for some other reason.

Reply:
Updated.